File: src/main/java/com/uiptv/service/DatabaseSyncService.java
package com.uiptv.service;

import com.uiptv.db.DatabaseUtils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.sql.SQLException;

import static com.uiptv.util.SQLiteTableSync.ensureDatabaseReady;
import static com.uiptv.util.SQLiteTableSync.replaceTable;
import static com.uiptv.util.SQLiteTableSync.syncPublishedM3uSelections;
import static com.uiptv.util.SQLiteTableSync.syncTables;

public class DatabaseSyncService {
    private static final List<DatabaseUtils.DbTable> CONFIGURATION_SYNCABLE = List.of(
            DatabaseUtils.DbTable.THEME_CSS_OVERRIDE_TABLE,
            DatabaseUtils.DbTable.PUBLISHED_M3U_SELECTION_TABLE
    );

    private DatabaseSyncService() {
    }

    private static class SingletonHelper {
        private static final DatabaseSyncService INSTANCE = new DatabaseSyncService();
    }

    public static DatabaseSyncService getInstance() {
        return SingletonHelper.INSTANCE;
    }

    public void syncDatabases(String sourceDB, String targetDB) throws SQLException {
        syncDatabases(sourceDB, targetDB, false, false);
    }

    public void syncDatabases(String sourceDB, String targetDB, boolean syncConfiguration, boolean syncExternalPlayerPaths) throws SQLException {
        syncDatabasesWithReport(sourceDB, targetDB, syncConfiguration, syncExternalPlayerPaths, null);
    }

    public DatabaseSyncReport syncDatabasesWithReport(String sourceDB,
                                                      String targetDB,
                                                      boolean syncConfiguration,
                                                      boolean syncExternalPlayerPaths,
                                                      SyncProgressListener progressListener) throws SQLException {
        ensureDatabaseReady(targetDB);
        List<TableSyncResult> tableResults = new ArrayList<>();
        int totalSteps = DatabaseUtils.Syncable.size() + (syncConfiguration ? 1 + CONFIGURATION_SYNCABLE.size() : 0);
        int completedSteps = 0;
        for (DatabaseUtils.DbTable tableName : DatabaseUtils.Syncable) {
            notifyProgress(progressListener, completedSteps, totalSteps, tableName.getTableName());
            int syncedRows = syncTables(sourceDB, targetDB, tableName);
            tableResults.add(new TableSyncResult(tableName.getTableName(), syncedRows));
            completedSteps++;
        }
        boolean configurationCopied = false;
        if (syncConfiguration) {
            notifyProgress(progressListener, completedSteps, totalSteps, DatabaseUtils.DbTable.CONFIGURATION_TABLE.getTableName());
            configurationCopied = com.uiptv.util.SQLiteTableSync.syncConfiguration(sourceDB, targetDB, syncExternalPlayerPaths);
            completedSteps++;
            notifyProgress(progressListener, completedSteps, totalSteps, DatabaseUtils.DbTable.THEME_CSS_OVERRIDE_TABLE.getTableName());
            tableResults.add(new TableSyncResult(
                    DatabaseUtils.DbTable.THEME_CSS_OVERRIDE_TABLE.getTableName(),
                    replaceTable(sourceDB, targetDB, DatabaseUtils.DbTable.THEME_CSS_OVERRIDE_TABLE)
            ));
            completedSteps++;

            notifyProgress(progressListener, completedSteps, totalSteps, DatabaseUtils.DbTable.PUBLISHED_M3U_SELECTION_TABLE.getTableName());
            tableResults.add(new TableSyncResult(
                    DatabaseUtils.DbTable.PUBLISHED_M3U_SELECTION_TABLE.getTableName(),
                    syncPublishedM3uSelections(sourceDB, targetDB)
            ));
            completedSteps++;
        }
        notifyProgress(progressListener, completedSteps, totalSteps, null);
        return new DatabaseSyncReport(tableResults, syncConfiguration, configurationCopied, syncExternalPlayerPaths);
    }

    private void notifyProgress(SyncProgressListener progressListener, int completedSteps, int totalSteps, String currentStep) {
        if (progressListener == null) {
            return;
        }
        progressListener.onProgress(completedSteps, totalSteps, currentStep);
    }

    @FunctionalInterface
    public interface SyncProgressListener {
        void onProgress(int completedSteps, int totalSteps, String currentStep);
    }

    public static final class DatabaseSyncReport {
        private final List<TableSyncResult> tableResults;
        private final boolean configurationRequested;
        private final boolean configurationCopied;
        private final boolean externalPlayerPathsIncluded;

        public DatabaseSyncReport(List<TableSyncResult> tableResults,
                                  boolean configurationRequested,
                                  boolean configurationCopied,
                                  boolean externalPlayerPathsIncluded) {
            this.tableResults = Collections.unmodifiableList(new ArrayList<>(tableResults));
            this.configurationRequested = configurationRequested;
            this.configurationCopied = configurationCopied;
            this.externalPlayerPathsIncluded = externalPlayerPathsIncluded;
        }

        public List<TableSyncResult> getTableResults() {
            return tableResults;
        }

        public boolean isConfigurationRequested() {
            return configurationRequested;
        }

        public boolean isConfigurationCopied() {
            return configurationCopied;
        }

        public boolean isExternalPlayerPathsIncluded() {
            return externalPlayerPathsIncluded;
        }

        public int getTotalRowsSynced() {
            return tableResults.stream().mapToInt(TableSyncResult::getRowCount).sum();
        }
    }

    public static final class TableSyncResult {
        private final String tableName;
        private final int rowCount;

        public TableSyncResult(String tableName, int rowCount) {
            this.tableName = tableName;
            this.rowCount = rowCount;
        }

        public String getTableName() {
            return tableName;
        }

        public int getRowCount() {
            return rowCount;
        }
    }
}

File: server/src/main/kotlin/com/uiptv/service/DatabaseSyncService.kt
package com.uiptv.service

import com.uiptv.db.DatabaseUtils
import com.uiptv.util.SQLiteTableSync.ensureDatabaseReady
import com.uiptv.util.SQLiteTableSync.replaceTable
import com.uiptv.util.SQLiteTableSync.syncPublishedM3uCategorySelections
import com.uiptv.util.SQLiteTableSync.syncPublishedM3uChannelSelections
import com.uiptv.util.SQLiteTableSync.syncPublishedM3uSelections
import com.uiptv.util.SQLiteTableSync.syncTables
import java.sql.SQLException
import java.util.Collections

object DatabaseSyncService {
    private val CONFIGURATION_SYNCABLE = listOf(
        DatabaseUtils.DbTable.THEME_CSS_OVERRIDE_TABLE,
        DatabaseUtils.DbTable.PUBLISHED_M3U_SELECTION_TABLE,
        DatabaseUtils.DbTable.PUBLISHED_M3U_CATEGORY_SELECTION_TABLE,
        DatabaseUtils.DbTable.PUBLISHED_M3U_CHANNEL_SELECTION_TABLE
    )


    @JvmStatic
    @Throws(SQLException::class)
    fun syncDatabases(sourceDB: String, targetDB: String) {
        syncDatabases(sourceDB, targetDB, false, false)
    }

    @JvmStatic
    @Throws(SQLException::class)
    fun syncDatabases(sourceDB: String, targetDB: String, syncConfiguration: Boolean, syncExternalPlayerPaths: Boolean) {
        syncDatabasesWithReport(sourceDB, targetDB, syncConfiguration, syncExternalPlayerPaths, null)
    }

    @JvmStatic
    @Throws(SQLException::class)
    fun syncDatabasesWithReport(
        sourceDB: String,
        targetDB: String,
        syncConfiguration: Boolean,
        syncExternalPlayerPaths: Boolean,
        progressListener: SyncProgressListener?
    ): DatabaseSyncReport {
        ensureDatabaseReady(targetDB)
        val tableResults = ArrayList<TableSyncResult>()
        val totalSteps = DatabaseUtils.Syncable.size + if (syncConfiguration) 1 + CONFIGURATION_SYNCABLE.size else 0
        var completedSteps = 0
        for (tableName in DatabaseUtils.Syncable) {
            notifyProgress(progressListener, completedSteps, totalSteps, tableName.tableName)
            val syncedRows = syncTables(sourceDB, targetDB, tableName)
            tableResults.add(TableSyncResult(tableName.tableName, syncedRows))
            completedSteps++
        }
        var configurationCopied = false
        if (syncConfiguration) {
            notifyProgress(progressListener, completedSteps, totalSteps, DatabaseUtils.DbTable.CONFIGURATION_TABLE.tableName)
            configurationCopied = com.uiptv.util.SQLiteTableSync.syncConfiguration(sourceDB, targetDB, syncExternalPlayerPaths)
            completedSteps++

            notifyProgress(progressListener, completedSteps, totalSteps, DatabaseUtils.DbTable.THEME_CSS_OVERRIDE_TABLE.tableName)
            tableResults.add(
                TableSyncResult(
                    DatabaseUtils.DbTable.THEME_CSS_OVERRIDE_TABLE.tableName,
                    replaceTable(sourceDB, targetDB, DatabaseUtils.DbTable.THEME_CSS_OVERRIDE_TABLE)
                )
            )
            completedSteps++

            notifyProgress(progressListener, completedSteps, totalSteps, DatabaseUtils.DbTable.PUBLISHED_M3U_SELECTION_TABLE.tableName)
            tableResults.add(
                TableSyncResult(
                    DatabaseUtils.DbTable.PUBLISHED_M3U_SELECTION_TABLE.tableName,
                    syncPublishedM3uSelections(sourceDB, targetDB)
                )
            )
            completedSteps++

            notifyProgress(progressListener, completedSteps, totalSteps, DatabaseUtils.DbTable.PUBLISHED_M3U_CATEGORY_SELECTION_TABLE.tableName)
            tableResults.add(
                TableSyncResult(
                    DatabaseUtils.DbTable.PUBLISHED_M3U_CATEGORY_SELECTION_TABLE.tableName,
                    syncPublishedM3uCategorySelections(sourceDB, targetDB)
                )
            )
            completedSteps++

            notifyProgress(progressListener, completedSteps, totalSteps, DatabaseUtils.DbTable.PUBLISHED_M3U_CHANNEL_SELECTION_TABLE.tableName)
            tableResults.add(
                TableSyncResult(
                    DatabaseUtils.DbTable.PUBLISHED_M3U_CHANNEL_SELECTION_TABLE.tableName,
                    syncPublishedM3uChannelSelections(sourceDB, targetDB)
                )
            )
            completedSteps++
        }
        notifyProgress(progressListener, completedSteps, totalSteps, null)
        return DatabaseSyncReport(tableResults, syncConfiguration, configurationCopied, syncExternalPlayerPaths)
    }

    private fun notifyProgress(progressListener: SyncProgressListener?, completedSteps: Int, totalSteps: Int, currentStep: String?) {
        progressListener?.onProgress(completedSteps, totalSteps, currentStep)
    }

    fun interface SyncProgressListener {
        fun onProgress(completedSteps: Int, totalSteps: Int, currentStep: String?)
    }

    class DatabaseSyncReport(
        tableResults: List<TableSyncResult>,
        val configurationRequested: Boolean,
        val configurationCopied: Boolean,
        val externalPlayerPathsIncluded: Boolean
    ) {
        val tableResults: List<TableSyncResult> = Collections.unmodifiableList(ArrayList(tableResults))

        fun tableResults(): List<TableSyncResult> = tableResults

        fun isConfigurationRequested(): Boolean = configurationRequested

        fun isConfigurationCopied(): Boolean = configurationCopied

        fun isExternalPlayerPathsIncluded(): Boolean = externalPlayerPathsIncluded

        fun getTotalRowsSynced(): Int = tableResults.sumOf { it.rowCount }
    }

    data class TableSyncResult(
        val tableName: String,
        val rowCount: Int
    ) {
        fun tableName(): String = tableName

        fun rowCount(): Int = rowCount
    }
}
